### PR TITLE
Issue-221: go back to lazy loading of definition files

### PR DESCRIPTION
**Pull request by ptrimble**
Fixes https://github.com/holidays/holidays/issues/221

Also fixes https://github.com/holidays/holidays/issues/222

(I heard from a coworker that if you reference the issue(s) in the PR description GH will automatically close them. Let's see if that's true!)

### CODE REVIEW

**Comment by ptrimble** [comment]
Hm. Why is this failing now? I understand that it's trying to use jruby 1.7 and we don't support ruby 1.9.3 but...that hasn't changed with this PR. Why is it breaking?

@ttwo32 What do you think about doing a major version bump to drop jruby 1.7 support? It's been about 5 months since the last major version bump. I'm tired of dealing with jruby 1.7, people need to update. Thoughts? Is 5 months too soon between major version bumps?

I admit that I am very aggressive with semver major version bumps. 😄

**Comment by ttwo32** [comment]
@ptrimble I agree with you.(I'm tired too.)
Anyway JRuby Dev team will stop support 1.7 in 2016. Everybody need to conevert to JRuby 9K from JRuby 1.7.

Go ahead and bump up.:smile:

**Comment by ptrimble** [comment]
Cool. I'm wondering if I should just...merge in my planned definitions changes with this? Just do it all at once?

I think I'll give it a shot and see if I like it. Let's see how the rest of this week goes. 😄

**Comment by ptrimble** [comment]
I am reasonably sure that no coverage has actually changed but since I have dropped/removed/changed file locations it is thinking something has changed. I'm gonna ignore it. 0.005 is not significant.

**Comment by ptrimble** [comment]
@ttwo32 I think I am done. Once we merge https://github.com/holidays/definitions/pull/1 I think we can then merge this. 

Can I ask a favor? Can you please read the CONTRIBUTING guide to make sure it makes sense? Also, can you help test that when you clone the repository fresh it correctly pulls the submodule when running tests? That's my biggest concern.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -6,25 +6,30 @@ There are multiple ways to help! We rely on users around the world to help keep
...
+## General note
+
+The definitions for this project are housed in a submodule. Please make sure to run `git clone --recursive git@github.com/holidays/holidays`
+or, if you forgot to do so, run `make update_defs` so that all of the submodule data is pulled down.
+
 ## For definition updates
 
-Our definitions are written in YAML. You can find a complete guide to our format in the [definitions README](definitions/README.md). We take the YAML definitions and generate final ruby classes that are loaded at runtime for fast calculations.
+Our definitions are written in YAML. They are housed in a [separate repository](https://github.com/holidays/definitions) so
+that they can be used by tools written in other languages. You can find a complete guide to our format in the
```
Comment (ttwo32) [inline CONTRIBUTING.md]: 👍

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -6,25 +6,30 @@ There are multiple ways to help! We rely on users around the world to help keep
 
 Please read our [Code of Conduct](CODE_OF_CONDUCT.md) before contributing. Everyone interacting with this project (or associated projects) is expected to abide by its terms.
 
+## General note
+
+The definitions for this project are housed in a submodule. Please make sure to run `git clone --recursive git@github.com/holidays/holidays`
```
Comment (ttwo32) [inline CONTRIBUTING.md]: Correct is 'git clone --recursive git@github.com:holidays/holidays' .:smile:

Comment (ttwo32) [inline CONTRIBUTING.md]: @ptrimble When i did 'make update_defs', but submodules didn't checkout.
I checked [help](https://git-scm.com/docs/git-submodule) and executed  'git submodule update --init', submodule was checked out.
So 'git submodule update --init' is needed in makefile.

Comment (ptrimble) [inline CONTRIBUTING.md]: I slightly modified the makefile target to use `git submodule update --remote --recursive`, which I found does the update I expected. Can you give it a shot yourself and see if it works @ttwo32?

Comment (ptrimble) [inline CONTRIBUTING.md]: Nope, wait, you are right. I think we need the initial init? But then after it is initialized the command I have above works. Hm....

Comment (ptrimble) [inline CONTRIBUTING.md]: Okay, so I've added `--init` to the `make update_defs` target and it seems to give us what we want. If you do the initial `git clone --recursive` then you will get the submodule as expected, but if you switch to another branch then you have to do the `make update_defs` with the new `--init` in order to get the latest. I tested this with a brand-new clone and I also verified that running with `--init` more than once doesn't cause issues.

Think we are good! Can you confirm @ttwo32?

Addressed in https://github.com/holidays/holidays/pull/222/commits/c52b13c553ac9ca2c0a0d8e0dbe9ccb1481ae1ce

Comment (ttwo32) [inline CONTRIBUTING.md]: @ptrimble I have retried `make update_def` command.It's OK!! Please go ahead.:D

**Comment by ptrimble** [comment]
Thanks @ttwo32! ❤️ 

I will see if I can get this merged and released tonight but it might wait until this weekend.

**Comment by ttwo32** [comment]
It's no problem at all.:D
I appreciate your hard work.:smile:
Thanks!!

**Comment by ptrimble** [comment]
Sorry for destroying history here but put myself into a bind because I'm so bad at git and decided to just start over. 👎 

Nothing was lost and we are back to where we were, I just rewrote the history on my branch a bit. I'll be merging this shortly if everything passes.

**Pull request merged**